crates/wasm-smith/src/lib.rs

```diff
@@ -1042,6 +1057,161 @@ where
+            let mut imports = Vec::with_capacity(module.num_imports);
+            for (module, name, ty) in module
+                .initial_sections
+                .iter()
+                .filter_map(|section| match section {
+                    InitialSection::Import(list) => Some(list),
+                    _ => None,
+                })
+                .flat_map(|a| a)
+            {
+                imports.push((module.clone(), name.clone(), ty.clone()));
+            }
+            let mut exports = Vec::with_capacity(module.exports.len());
+            for (name, item) in module.exports.iter() {
+                let ty = module.type_of(item);
+                exports.push((name.clone(), ty));
+            }
+            let ty = Rc::new(ModuleType {
+                imports,
+                exports: Rc::new(InstanceType { exports }),
+            });
+
+            // And then given the type of the module we copy it over to
+            // ourselves and get a type index. This index goes into the module
+            // section and we also record the type of this module in our
+            // `modules` array.
+            let ty_idx = self.get_module_type_idx(&ty);
+            modules.push(ty_idx);
+            self.modules.push(self.module_type(ty_idx).clone());
+            self.defined_modules.push(module);
+            Ok(true)
+        })?;
+        if !modules.is_empty() || u.arbitrary()? {
+            self.initial_sections.push(InitialSection::Module(modules));
+        }
+        Ok(())
+    }
+
+    /// Copies the type `ty`, possibly from a different module, into this
+    /// module.
+    fn copy_type_to_self(&mut self, ty: &EntityType) -> EntityType {
+        match ty {
+            // Globals, tables, and memories are easy, it's just copying some
+            // bits around.
+            EntityType::Global(t) => EntityType::Global(t.clone()),
+            EntityType::Table(t) => EntityType::Table(t.clone()),
+            EntityType::Memory(t) => EntityType::Memory(t.clone()),
+
+            // Functions, instances, and modules are hairier, however, since
+            // they chases pointers. We delegate to some helper methods for
+            // each one.
+            EntityType::Func(_other_idx, sig) => {
+                EntityType::Func(self.get_func_type_idx(sig), sig.clone())
+            }
+            EntityType::Instance(_other_idx, sig) => {
+                EntityType::Instance(self.get_instance_type_idx(sig), sig.clone())
+            }
+            EntityType::Module(_other_idx, sig) => {
+                EntityType::Module(self.get_module_type_idx(sig), sig.clone())
+            }
+        }
+    }
+
+    fn get_func_type_idx(&mut self, sig: &Rc<FuncType>) -> u32 {
+        // First check to see if our module already defines this function
+        // signature...
+        if let Some((i, _)) = self.func_types().find(|(_, ty)| **ty == **sig) {
+            return i;
+        }
```

Comment:
We may want to make a `HashMap<Rc<FuncType>>, u32>`, although then I suppose we would never generate modules with duplicate function type definitions.

But also, as written, this will always use the first occurrence of a given function type if there are duplicates. We *could* avoid this by starting this search at a random index `i` and checking each type `j` in the range `i..i+self.num_func_types` via `self.func_types()[j % self.num_func_types]`. But this doesn't really seem worth the hassle.

Anyways, I think the hash consing is probably a good thing to do, rather than linear search, since actually doing duplicate function types well is a bit of a pain and also not very useful.

Reply:
I'll go ahead and implement `HashMap<Rc<FuncType>, Vec<u32>>` and that should be easy enough to implement, and it makes all possible function signatures a candidate too!